src/couch_file.erl

```diff
@@ -531,27 +537,96 @@ find_header(Fd, Block) ->
+        <<RawBin:TotalSize/binary, _/binary>> = RestBlock;
     true ->
         {ok, Missing} = file:pread(
             Fd, (Block * ?SIZE_BLOCK) + 5 + byte_size(RestBlock),
-            TotalBytes - byte_size(RestBlock)),
+            TotalSize - byte_size(RestBlock)),
         RawBin = <<RestBlock/binary, Missing/binary>>
     end,
     <<Md5Sig:16/binary, HeaderBin/binary>> =
         iolist_to_binary(remove_block_prefixes(5, RawBin)),
     Md5Sig = couch_crypto:hash(md5, HeaderBin),
     {ok, HeaderBin}.
 
+
+%% Read a configurable number of block locations at a time using
+%% file:pread/2.  At each block location, read ?PREFIX_SIZE (5) bytes;
+%% if the first byte is <<1>>, assume it's a header block, and the
+%% next 4 bytes hold the size of the header.
+-spec find_header_pread2(file:fd(), block_id()) ->
+    {ok, binary()} | no_valid_header.
+find_header_pread2(Fd, Block) ->
+    find_header_pread2(Fd, Block, read_count()).
+
+-spec find_header_pread2(file:fd(), block_id(), non_neg_integer()) ->
+    {ok, binary()} | no_valid_header.
+find_header_pread2(_Fd, Block, _ReadCount) when Block < 0 ->
+    no_valid_header;
+find_header_pread2(Fd, Block, ReadCount) ->
+    Locations = block_locations(Block, ReadCount),
+    {ok, DataL} = file:pread(Fd, [{L, ?PREFIX_SIZE} || L <- Locations]),
+    case locate_header(Fd, header_locations(Locations, DataL)) of
+        {ok, _Location, HeaderBin} ->
+            io:format("header at block ~p~n", [_Location div ?SIZE_BLOCK]), % TEMP
+            {ok, HeaderBin};
+        _ ->
+            ok = file:advise(
+                Fd, hd(Locations), ReadCount * ?SIZE_BLOCK, dont_need),
+            NextBlock = hd(Locations) div ?SIZE_BLOCK - 1,
+            find_header_pread2(Fd, NextBlock, ReadCount)
+    end.
+
+-spec read_count() -> non_neg_integer().
+read_count() ->
+    config:get_integer("couchdb", "find_header_read_count", ?DEFAULT_READ_COUNT).
+
+-spec block_locations(block_id(), non_neg_integer()) -> [location()].
+block_locations(Block, ReadCount) ->
+    First = max(0, Block - ReadCount + 1),
+    [?SIZE_BLOCK*B || B <- lists:seq(First, Block)].
+
+-spec header_locations([location()], [data()]) -> [{location(), header_size()}].
+header_locations(Locations, DataL) ->
+    lists:foldl(fun
+        ({Loc, <<1, HeaderSize:32/integer>>}, Acc) ->
+            [{Loc, HeaderSize} | Acc];
+        (_, Acc) ->
+            Acc
+        end, [], lists:zip(Locations, DataL)).
+
+-spec locate_header(file:fd(), [{location(), header_size()}]) ->
+    {ok, binary()} | not_found.
+locate_header(_Fd, []) ->
+    not_found;
+locate_header(Fd, [{Location, Size}|LocationSizes]) ->
+    case (catch load_header(Fd, Location, Size)) of
+        {ok, HeaderBin} ->
+            {ok, Location, HeaderBin};
+        _Error ->
+            locate_header(Fd, LocationSizes)
+    end.
+
+-spec load_header(file:fd(), location(), header_size()) ->
+    {ok, binary()} | no_return().
+load_header(Fd, Location, Size) ->
```

Comment:
You should move this function up under load_header/2 and then rearrange load_header/2 so that it calls this function. Scrolling back and forth between the two trying to figure out the difference was quite confusing.